internal/mesh/internal/controllers/xds/controller.go

```diff
@@ -174,11 +310,93 @@ func (r *xdsReconciler) Reconcile(ctx context.Context, rt controller.Runtime, re
+// bimapper package uses resource references, so we use an internal type to create a leaf resource reference since leaf
+// certificates are not v2 resources.
+func leafResourceRef(workloadIdentity, namespace, partition string) *pbresource.Reference {
+	// Since leaf certificate references aren't resources in the resource API, we don't have the same guarantees that
+	// namespace and partition are set. So this is to ensure that we always do set values for tenancy.
+	if namespace == "" {
+		namespace = defaultTenancy
+	}
+	if partition == "" {
+		partition = defaultTenancy
+	}
 	ref := &pbresource.Reference{
-		Name:    id.GetName(),
-		Type:    id.GetType(),
-		Tenancy: id.GetTenancy(),
+		Name: workloadIdentity,
+		Type: InternalLeafType,
+		Tenancy: &pbresource.Tenancy{
+			Partition: partition,
+			Namespace: namespace,
+		},
 	}
 	return ref
 }
+
+// InternalLeafType sets up an internal resource type to use for leaf certificates, since they are not yet a v2
+// resource. It's exported because it's used by the mesh controller registration which needs to set up the bimapper for
+// leaf certificates.
+var InternalLeafType = &pbresource.Type{
+	Group:        "internal",
+	GroupVersion: "v1alpha1",
+	Kind:         "leaf",
+}
+
+// keyFromReference is used to create string keys from resource references.
+func keyFromReference(ref resource.ReferenceOrID) string {
+	return fmt.Sprintf("%s/%s/%s",
+		resource.ToGVK(ref.GetType()),
+		tenancyToString(ref.GetTenancy()),
+		ref.GetName())
+}
+
+func tenancyToString(tenancy *pbresource.Tenancy) string {
+	return fmt.Sprintf("%s.%s", tenancy.Partition, tenancy.Namespace)
+}
+
+// generateProxyStateLeafCertificates translates a *structs.IssuedCert into a *pbproxystate.LeafCertificate.
+func generateProxyStateLeafCertificates(cert *structs.IssuedCert) *pbproxystate.LeafCertificate {
+	if cert.CertPEM == "" || cert.PrivateKeyPEM == "" {
+		return nil
+	}
+	return &pbproxystate.LeafCertificate{
+		Cert: cert.CertPEM,
+		Key:  cert.PrivateKeyPEM,
+	}
+}
+
+// cancelWatches cancels watches for leafs that no longer need to be watched, as long as it is referenced by zero ProxyStateTemplates.
+func (r *xdsReconciler) cancelWatches(leafResourceRefs []*pbresource.Reference) {
+	for _, leaf := range leafResourceRefs {
+		pstItems := r.leafMapper.ItemRefsForLink(leaf)
+		if len(pstItems) > 0 {
+			// Don't delete and cancel watches, since this leaf is referenced elsewhere.
+			continue
+		}
+		cancel, ok := r.leafCancels.Get(keyFromReference(leaf))
+		if ok {
+			cancel()
+			r.leafCancels.Delete(keyFromReference(leaf))
+		}
+	}
+}
+
+// prevWatchesToCancel computes if there are any items in prevWatchedLeafs that are not in currentLeafs, and returns a list of those items.
+func prevWatchesToCancel(prevWatchedLeafs []*pbresource.Reference, currentLeafs []resource.ReferenceOrID) []*pbresource.Reference {
+	var prevWatchedLeafsToCancel []*pbresource.Reference
+	for _, prevLeaf := range prevWatchedLeafs {
```

Comment:
Nit. This can likely be addressed in a TODO later if need be, but this is an `N*M` operation, when it could be faster by using map lookups. When we are doing lots of reconciles, this sort of nested loop could be an issue.